src/stream/src/executor/dynamic_filter.rs

```diff
@@ -41,7 +44,210 @@ use super::{
+        match (
+            self.get_cached_value(),
+            row.datum_at(self.watermark_col_idx),
+        ) {
+            (None, Some(inserted_value)) => {
+                self.entry = DynamicFilterWatermarkCacheEntry::Smallest {
+                    pk: (&row).project(&self.lhs_pk_indices).into_owned_row(),
+                    value: inserted_value.into_scalar_impl(),
+                }
+            }
+            (Some(current_value), Some(inserted_value)) => {
+                if inserted_value
+                    .default_cmp(&current_value.as_scalar_ref_impl())
+                    .is_lt()
+                {
+                    self.entry = DynamicFilterWatermarkCacheEntry::Smallest {
+                        pk: (&row).project(&self.lhs_pk_indices).into_owned_row(),
+                        value: inserted_value.into_scalar_impl(),
+                    }
+                }
+            }
+            _ => {}
+        }
+    }
+
+    fn process_delete(&mut self, row: impl Row) {
+        if let Some((pk, _val)) = self.get_cached_pk_and_val() {
+            let del_pk = row.project(&self.lhs_pk_indices);
+            if Row::eq(pk, del_pk) {
+                self.entry = DynamicFilterWatermarkCacheEntry::Evicted;
+            }
+        }
+    }
+
+    /// Takes care of state table commit
+    /// Trade-off the cost if we have Delete-heavy workload.
+    /// Instead of issuing delete range on every barrier,
+    /// we will do table scan instead, which seems more acceptable.
+    async fn process_state_table_commit<S: StateStore>(
+        &mut self,
+        left_table: &StateTable<S>,
+    ) -> StreamExecutorResult<()> {
+        let watermark_opt = left_table.get_state_clean_watermark();
+        if let Some(watermark) = watermark_opt {
+            self.prev_cleaned_watermark = Some(watermark.clone());
+            if self.entry == DynamicFilterWatermarkCacheEntry::Evicted {
+                let range: (Bound<Once<Datum>>, Bound<Once<Datum>>) =
+                    (Included(once(Some(watermark.clone()))), Unbounded);
+                let mut streams = vec![];
+                for vnode in left_table.vnodes().iter_vnodes() {
+                    // TODO(kwannoel): Is it possible to have a state_table interface to read the
+                    // 1st value per vnode? Since that's all we need.
+                    let stream = left_table
+                        .iter_key_and_val_with_pk_range(
+                            &range,
+                            vnode,
+                            PrefetchOptions::new_for_exhaust_iter(),
+                        )
+                        .await?;
+                    streams.push(Box::pin(stream));
+                }
+                let merged_stream = merge_sort(streams);
+                pin_mut!(merged_stream);
+                if let Some((_pk, row)) = merged_stream.next().await.transpose()? {
+                    self.process_insert(row);
+                }
```

Comment:
Refresh logic for cache.

Reply:
Btw can't use `merge_sort` directly. Merge sort operate on `pk`. We want to operate on watermark col.

Reply:
merge_sort works because individually, the streams coming from storage are ordered by pk, so it can just compare them by the pk.
But with ordering by watermark column, there’s no such guarantee.
It needs to change to scan all rows between prev_cleaned_watermark and current_watermark and order them by itself

Reply:
Unless watermark is pk prefix.

Reply:
can be used, watermark is pk prefix.